elm-src/index.elm

```diff
@@ -0,0 +1,489 @@
+    }
+
+
+update_page : Model -> Int -> Model
+update_page model newPage =
+    clamped
+        { model
+            | page = newPage
+            , selected = Nothing
+        }
+
+
+select_option : Model -> String -> Model
+select_option model opt =
+    { model
+        | selected = Just opt
+    }
+
+
+deselect_option : Model -> Model
+deselect_option model =
+    { model
+        | selected = Nothing
+    }
+
+
+init : Navigation.Location -> ( Model, Cmd Msg )
+init location =
+    let
+        qs =
+            QueryString.parse location.search
+
+        query =
+            Maybe.withDefault "" (QueryString.one QueryString.string "query" qs)
+
+        page =
+            Maybe.withDefault 1 (QueryString.one QueryString.int "page" qs)
+
+        selected =
+            QueryString.one QueryString.string "selected" qs
+    in
+        ( Model query (splitQuery query) [] [] page location selected Loading Debounce.init
+        , getOptions
+        )
+
+
+splitQuery : String -> List String
+splitQuery query =
+    String.words (String.toLower query)
+
+
+
+-- UPDATE
+
+
+updateFromUrl : Navigation.Location -> Model -> Model
+updateFromUrl location model =
+    let
+        qs =
+            QueryString.parse location.search
+
+        query =
+            Maybe.withDefault "" (QueryString.one QueryString.string "query" qs)
+
+        page =
+            Maybe.withDefault 1 (QueryString.one QueryString.int "page" qs)
+
+        selected =
+            QueryString.one QueryString.string "selected" qs
+    in
+        case selected of
+            Just opt ->
+                select_option model opt
+
+            Nothing ->
+                (update_page (update_query model query) page)
+
+
+updateUrl : Model -> Cmd Msg
+updateUrl model =
+    let
+        querystring =
+            case model.selected of
+                Nothing ->
+                    String.append
+                        (String.append "?query=" (Http.encodeUri model.query))
+                        (if model.page > 1 then
+                            (String.append "&page=" (Http.encodeUri (toString model.page)))
+                         else
+                            ""
+                        )
+
+                Just opt ->
+                    (String.append "?selected=" (Http.encodeUri opt))
+    in
+        Navigation.newUrl querystring
+
+
+type Msg
+    = ChangeQuery String
+    | ChangePage Int
+    | ChangeUrl Navigation.Location
+    | FetchedOptions (Result Http.Error NixOptions)
+    | SelectOption String
+    | DeselectOption
+    | Bounce (Debounce.Msg Msg)
+    | SyncQueryToUrl
+
+
+type PageStatus
+    = Loading
+    | Loaded
+    | Error String
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        FetchedOptions (Ok options) ->
+            let
+                newModel =
+                    update_options model options
+            in
+                ( newModel, updateUrl newModel )
```

Comment:
Perhaps an alternative is to rewrite `updateUrl` so it's type-signature can be `updateUrl : Model -> (Model, Cmd Msg)`. If you do that, then these branches can be written as `model |> update_options options |> updateUrl`. (you'd have to flip the params for `updateOptions`, but in general putting the "core" datastrcture a function works on as the last argument is good form, since it's pipeline-friendly)